File: android/src/main/kotlin/com/neelakandan/flutter_neo_shield/rasp/NetworkThreatDetector.kt
package com.neelakandan.flutter_neo_shield.rasp

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.os.Build
import java.net.NetworkInterface

/**
 * Detects network-level threats used during APK reverse engineering from desktop:
 *
 * 1. HTTP Proxy — Burp Suite, mitmproxy, Charles Proxy running on desktop
 * 2. VPN — VPN tunnels routing traffic through desktop interceptors
 *
 * These are the primary tools attackers use to intercept HTTPS traffic
 * after decompiling and repackaging an APK.
 */
class NetworkThreatDetector {

    /**
     * Returns a map with:
     *   "proxyDetected" -> Boolean
     *   "vpnDetected"   -> Boolean
     *   "detected"      -> Boolean (true if ANY threat found)
     */
    fun check(context: Context): Map<String, Any> {
        val proxyDetected = checkProxy(context)
        val vpnDetected = checkVpn(context)

        return mapOf(
            "proxyDetected" to proxyDetected,
            "vpnDetected" to vpnDetected,
            "detected" to (proxyDetected || vpnDetected)
        )
    }

    /**
     * Simple boolean: returns true if proxy or VPN detected.
     */
    fun checkSimple(context: Context): Boolean {
        return checkProxy(context) || checkVpn(context)
    }

    /**
     * Detects HTTP/HTTPS proxy configuration.
     *
     * Attackers configure proxy settings on the device/emulator to route
     * traffic through Burp Suite or mitmproxy on their desktop.
     */
    private fun checkProxy(context: Context): Boolean {
        // Method 1: System property check
        try {
            val httpProxy = System.getProperty("http.proxyHost")
            if (!httpProxy.isNullOrEmpty()) {
                return true
            }

            val httpsProxy = System.getProperty("https.proxyHost")
            if (!httpsProxy.isNullOrEmpty()) {
                return true
            }
        } catch (e: Exception) {
            // Ignore
        }

        // Method 2: ConnectivityManager proxy info (API 23+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            try {
                val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
                val network = cm?.activeNetwork
                val linkProperties = cm?.getLinkProperties(network)
                val proxyInfo = linkProperties?.httpProxy
                if (proxyInfo != null && proxyInfo.host != null) {
                    return true
                }
            } catch (e: Exception) {
                // Ignore
            }
        }

        // Method 3: Global proxy setting
        try {
            val globalProxy = android.provider.Settings.Global.getString(
                context.contentResolver,
                android.provider.Settings.Global.HTTP_PROXY
            )
            if (!globalProxy.isNullOrEmpty() && globalProxy != ":0") {
                return true
            }
        } catch (e: Exception) {
            // Ignore
        }

        return false
    }

    /**
     * Detects active VPN connections.
     *
     * VPN tunnels are used to route all device traffic through a desktop
     * machine running an intercepting proxy.
     */
    private fun checkVpn(context: Context): Boolean {
        // Method 1: ConnectivityManager network capabilities (API 23+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            try {
                val cm = context.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
                val activeNetwork = cm?.activeNetwork
                if (activeNetwork != null) {
                    val caps = cm.getNetworkCapabilities(activeNetwork)
                    if (caps != null && caps.hasTransport(NetworkCapabilities.TRANSPORT_VPN)) {
                        return true
                    }
                }
            } catch (e: Exception) {
                // Ignore
            }
        }

        // Method 2: Check for tun/ppp/tap network interfaces
        try {
            val interfaces = NetworkInterface.getNetworkInterfaces()
            while (interfaces.hasMoreElements()) {
                val iface = interfaces.nextElement()
                if (!iface.isUp) continue
                val name = iface.name.lowercase()
                if (name.startsWith("tun") ||
                    name.startsWith("ppp") ||
                    name.startsWith("tap") ||
                    name.startsWith("utun") ||
                    name.startsWith("ipsec")) {
                    return true
                }
            }
        } catch (e: Exception) {
            // Ignore
        }

        return false
    }
}

File: ios/Classes/rasp/NetworkThreatDetector.swift
import Foundation
import Darwin

/// Detects network-level threats used during reverse engineering from desktop:
///
/// 1. HTTP Proxy — Burp Suite, mitmproxy, Charles Proxy
/// 2. VPN — VPN tunnels routing traffic through interceptors
///
/// These detect MITM (Man-in-the-Middle) attack setups commonly used
/// alongside APK/IPA reverse engineering.
public class NetworkThreatDetector {

    /// Returns a dictionary with detection results:
    ///   "proxyDetected" -> Bool
    ///   "vpnDetected"   -> Bool
    ///   "detected"      -> Bool (true if ANY threat)
    public static func check() -> [String: Any] {
        let proxyDetected = checkProxy()
        let vpnDetected = checkVpn()

        return [
            "proxyDetected": proxyDetected,
            "vpnDetected": vpnDetected,
            "detected": proxyDetected || vpnDetected
        ]
    }

    /// Simple boolean: true if proxy or VPN detected.
    public static func checkSimple() -> Bool {
        return checkProxy() || checkVpn()
    }

    /// Detects HTTP/HTTPS proxy configuration.
    ///
    /// Attackers configure proxy on the device to route traffic through
    /// Burp Suite, mitmproxy, or Charles Proxy on their desktop.
    private static func checkProxy() -> Bool {
        // Method 1: CFNetwork proxy settings
        guard let proxySettings = CFNetworkCopySystemProxySettings()?.takeRetainedValue() as? [String: Any] else {
            return false
        }

        // Check HTTP proxy
        if let httpProxy = proxySettings[kCFNetworkProxiesHTTPProxy as String] as? String,
           !httpProxy.isEmpty {
            if let httpEnabled = proxySettings[kCFNetworkProxiesHTTPEnable as String] as? Int,
               httpEnabled == 1 {
                return true
            }
        }

        // Check HTTPS proxy
        if let httpsProxy = proxySettings[kCFNetworkProxiesHTTPSProxy as String] as? String,
           !httpsProxy.isEmpty {
            if let httpsEnabled = proxySettings[kCFNetworkProxiesHTTPSEnable as String] as? Int,
               httpsEnabled == 1 {
                return true
            }
        }

        // Check SOCKS proxy (used by some tools)
        if let socksProxy = proxySettings[kCFNetworkProxiesSOCKSProxy as String] as? String,
           !socksProxy.isEmpty {
            if let socksEnabled = proxySettings[kCFNetworkProxiesSOCKSEnable as String] as? Int,
               socksEnabled == 1 {
                return true
            }
        }

        return false
    }

    /// Detects active VPN connections.
    ///
    /// VPN tunnels route all traffic through a desktop interceptor.
    private static func checkVpn() -> Bool {
        // Check network interfaces for VPN tunnels
        var ifaddr: UnsafeMutablePointer<ifaddrs>?
        guard getifaddrs(&ifaddr) == 0, let firstAddr = ifaddr else {
            return false
        }
        defer { freeifaddrs(ifaddr) }

        let vpnPrefixes = ["utun", "ppp", "ipsec", "tap", "tun"]

        var addr = firstAddr
        while true {
            let name = String(cString: addr.pointee.ifa_name)
            let flags = Int32(addr.pointee.ifa_flags)
            let isUp = (flags & IFF_UP) != 0

            if isUp {
                for prefix in vpnPrefixes {
                    if name.hasPrefix(prefix) {
                        return true
                    }
                }
            }

            guard let next = addr.pointee.ifa_next else { break }
            addr = next
        }

        return false
    }
}
